File: api/Controllers/BaseController.cs
using System.Diagnostics;
using System.Security.Claims;
using Football_Management.API.Models.Entities.Base;
using Football_Management.API.Services.Interfaces;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;

namespace Football_Management.API.Controllers.Base;

[ApiController]
public abstract class BaseController<TEntity, TResponse, TCreateRequest, TUpdateRequest>
    : ControllerBase
    where TEntity : BaseEntity
{
    protected readonly IBaseService<TEntity, TResponse, TCreateRequest, TUpdateRequest> Service;
    protected readonly ILogger Logger;

    protected BaseController(
        IBaseService<TEntity, TResponse, TCreateRequest, TUpdateRequest> service,
        ILogger logger
        )
    {
        Service = service;
        Logger = logger;
    }

    // ── CRUD ──────────────────────────────────────────────────────
    [Authorize]
    [HttpGet]
    public virtual async Task<IActionResult> GetAll()
    {
        var sw = Stopwatch.StartNew();
        var result = await Service.GetAllAsync();
        Logger.LogInformation("{Entity} GetAll {Elapsed}ms", typeof(TEntity).Name, sw.ElapsedMilliseconds);
        return Ok(result);
    }

    [Authorize]
    [HttpGet("{id}")]
    public virtual async Task<IActionResult> GetById(int id)
    {
        var sw = Stopwatch.StartNew();
        var result = await Service.GetByIdAsync(id);
        Logger.LogInformation("{Entity} GetAll {Elapsed}ms", typeof(TEntity).Name, sw.ElapsedMilliseconds);
        return Ok(result);
    }
    [Authorize]
    [HttpPost]
    public virtual async Task<IActionResult> Create([FromBody] TCreateRequest request)
    {
        var sw = Stopwatch.StartNew();
        var result = await Service.CreateAsync(request);
        Logger.LogInformation("{Entity} Created {Elapsed}ms",
            typeof(TEntity).Name, sw.ElapsedMilliseconds);
        return StatusCode(201, result);
    }

    [Authorize]
    [HttpPut("{id}")]
    public virtual async Task<IActionResult> Update(int id, [FromBody] TUpdateRequest request)
    {
        var sw = Stopwatch.StartNew();
        var result = await Service.UpdateAsync(id, request);
        Logger.LogInformation("{Entity}#{Id} Updated {Elapsed}ms",
            typeof(TEntity).Name, id, sw.ElapsedMilliseconds);
        return Ok(result);
    }

    [Authorize]
    [HttpDelete("{id}")]
    public virtual async Task<IActionResult> Delete(int id)
    {
        var sw = Stopwatch.StartNew();
        await Service.DeleteAsync(id);
        Logger.LogInformation("{Entity}#{Id} Deleted {Elapsed}ms",
            typeof(TEntity).Name, id, sw.ElapsedMilliseconds);
        return NoContent();
    }

    // [HttpPatch("{id}/restore")]
    // public virtual async Task<IActionResult> Restore(int id)
    // {
    //     await Service.RestoreAsync(id);
    //     return NoContent();
    // }

    // [HttpDelete("{id}/permanent")]
    // public virtual async Task<IActionResult> DeletePermanently(int id)
    // {
    //     await Service.DeletePermanentlyAsync(id);
    //     return NoContent();
    // }
}

File: api/src/controllers/basecontroller.ts
import { Request, Response, NextFunction } from "express";
import { makeResponse } from "../common/response/apiresponse.js";
import { BaseService } from "../services/baseservice.js";

export abstract class BaseController<T, CreateDTO, UpdateDTO> {
    constructor(
        protected readonly service: BaseService<T, CreateDTO, UpdateDTO>
    ) { }

    protected ok<D>(res: Response, data: D, message = "Success"): void {
        res.status(200).json(makeResponse(data, message));
    }

    protected created<D>(res: Response, data: D, message = "Created"): void {
        res.status(201).json(makeResponse(data, message));
    }

    protected noContent(res: Response, message = "Deleted"): void {
        res.status(204).json(makeResponse(null, message));
    }

    findAll = async (_req: Request, res: Response, next: NextFunction) => {
        try {
            this.ok(res, await this.service.findAll());
        } catch (err) { next(err); }
    };

    findById = async (req: Request, res: Response, next: NextFunction) => {
        try {
            this.ok(res, await this.service.findByIdOrFail(Number(req.params.id)));
        } catch (err) { next(err); }
    };

    create = async (req: Request, res: Response, next: NextFunction) => {
        try {
            this.created(res, await this.service.create(req.body));
        } catch (err) { next(err); }
    };

    update = async (req: Request, res: Response, next: NextFunction) => {
        try {
            this.ok(res, await this.service.update(Number(req.params.id), req.body), "Updated");
        } catch (err) { next(err); }
    };

    softDelete = async (req: Request, res: Response, next: NextFunction) => {
        try {
            await this.service.softDelete(Number(req.params.id));
            this.noContent(res);
        } catch (err) { next(err); }
    };
}